TokenCacheMigration/CommonCacheAdalV5/FilesBasedTokenCache.cs

```diff
@@ -0,0 +1,119 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+using Microsoft.Identity.Core.Cache;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using System.IO;
+using System.Security.Cryptography;
+
+namespace CommonCacheADALV5
+{
+ /// <summary>
+ /// Simple file based persistent cache implementation for a desktop application (from ADAL 4.x)
+ /// </summary>
+ class FilesBasedTokenCache : TokenCache
+ {
+  public string AdalV3CacheFilePath { get; }
+  public string UnifiedCacheFilePath { get; }
+  private static readonly object FileLock = new object();
+
+  // Initializes the cache against a local file.
+  // If the file is already rpesent, it loads its content in the ADAL cache
+  public FilesBasedTokenCache(string adalV3FilePath, string unifiedCacheFilePath)
+  {
+   AdalV3CacheFilePath = adalV3FilePath;
+   AfterAccess = AfterAccessNotification;
+   UnifiedCacheFilePath = unifiedCacheFilePath;
+   BeforeAccess = BeforeAccessNotification;
+   BeforeAccessNotification(null);
+  }
+
+  // Empties the persistent store.
+  public override void Clear()
+  {
+   base.Clear();
+   File.Delete(AdalV3CacheFilePath);
+   File.Delete(UnifiedCacheFilePath);
+  }
+
+  // Triggered right before ADAL needs to access the cache.
+  // Reload the cache from the persistent store in case it changed since the last access.
+  void BeforeAccessNotification(TokenCacheNotificationArgs args)
+  {
+   lock (FileLock)
+   {
+    DeserializeAdalV3(ReadFromFileIfExists(AdalV3CacheFilePath));
+    DeserializeMsalV3(ReadFromFileIfExists(UnifiedCacheFilePath));
+   }
+  }
+
+  // Triggered right after ADAL accessed the cache.
+  void AfterAccessNotification(TokenCacheNotificationArgs args)
+  {
+   // if the access operation resulted in a cache update
+   if (this.HasStateChanged)
+   {
+    lock (FileLock)
+    {
+     // reflect changes in the persistent store
+     WriteToFileIfNotNull(AdalV3CacheFilePath, SerializeAdalV3());
+     WriteToFileIfNotNull(UnifiedCacheFilePath, SerializeMsalV3());
+
+     // once the write operation took place, restore the HasStateChanged bit to false
```

Comment:
does this compile? You don't need to set this flag.

Reply:
this does compile. This is ADAL here, @bgavrilMS !

Reply:
Ok :). It still does nothing, but we've been showing this pattern everywhere...

Reply:
it does nothing in ADAL as well?